text/0625-helper-managers.md

```diff
@@ -0,0 +1,504 @@
+
+`createHelper` is a required hook on the HelperManager interface. The hook is
+passed the definition of the helper that is currently being created, and is
+expected to return a _state bucket_. This state bucket is what represents the
+current state of the helper, and will be passed to the other lifecycle hooks at
+appropriate times. It is not necessarily related to the definition of the
+helper itself - for instance, you could return an object _containing_ an
+instance of the helper:
+
+```js
+class MyManager {
+  createHelper(Definition, args) {
+    return {
+      instance: new Definition(args);
+    };
+  }
+}
+```
+
+This allows the manager to store metadata that it doesn't want to expose to the
+user.
+
+This hook has the following timing semantics:
+
+**Always**
+- called as discovered during DOM construction
+- called in definition order in the template
+
+#### `getValue`
+
+`getValue` is an optional hook that should return the value of the helper. This
+is the value that is returned from the helper and passed into the template.
+
+This hook is called when the value is requested from the helper (e.g. when the
+template is rendering and the helper value is needed). The hook is autotracked,
+and will rerun whenever any tracked values used inside of it are updated.
+Otherwise it does not rerun.
+
+> Note: This means that arguments which are not _consumed_ within the hook will
+> not trigger updates.
+
+This hook is only called for helpers with the `hasValue` capability enabled.
+This hook has the following timing semantics:
+
+**Always**
+- called the first time the helper value is requested
+- called after autotracked state has changed
+
+**Never**
+- called if the `hasValue` capability is disabled
+
+#### `runEffect`
+
+`runEffect` is an optional hook that should run the effect that the helper is
+applying, setting it up or updating it.
+
+This hook is scheduled to be called after render. The hook is autotracked, and
+will rerun whenever any tracked values used inside of it are updated. Otherwise
+it does not rerun.
+
+The hook is also run during a time period where state mutations are _disabled_
+in Ember. Any change to any tracked property or tag via `Ember.set` will throw
+an error during this time. This is meant to prevent infinite rerenders and other
+antipatterns.
+
+This hook is only called for helpers with the `isScheduledEffect` capability
+enabled. It has the following timing semantics:
+
+**Always**
+- called after the helper was first created
+- called after autotracked state has changed
+
+**Never**
+- called if the `isScheduledEffect` capability is disabled
+
+#### `getDestroyable`
+
+`getDestroyable` is an optional hook that users can use to register a
+destroyable object for the helper. This destroyable will be registered to the
+containing block or template parent, and will be destroyed when it is destroyed.
+See the [Destroyables RFC](https://github.com/emberjs/rfcs/blob/master/text/0580-destroyables.md)
```

Comment:
It does exist in that RFC actually, though it could be easy to miss:

> For the remainder of this RFC, the terms "destroyable" and "destroyable object" will be used to mean any object which is a valid WeakMap key (e.g. typeof obj === 'object' || typeof obj === 'function'). Any JS object that fulfills this property can be used with this system.

Reply:
Aha! Thank you! 🤦